### ISSUE TITLE
‘O_NOFOLLOW’ undeclared (first use in this function)

### ISSUE THREAD

**Issue by george-b** [neovim/neovim#4042]
When pulling from master I get the following build error

```
/users/gbrown/src/binstrap/src/neovim/src/nvim/fileio.c: In function ‘buf_write’:
/users/gbrown/src/binstrap/src/neovim/src/nvim/fileio.c:2719: error: ‘O_NOFOLLOW’ undeclared (first use in this function)
/users/gbrown/src/binstrap/src/neovim/src/nvim/fileio.c:2719: error: (Each undeclared identifier is reported only once
/users/gbrown/src/binstrap/src/neovim/src/nvim/fileio.c:2719: error: for each function it appears in.)
/users/gbrown/src/binstrap/src/neovim/src/nvim/fileio.c: In function ‘vim_rename’:
/users/gbrown/src/binstrap/src/neovim/src/nvim/fileio.c:4569: error: ‘O_NOFOLLOW’ undeclared (first use in this function)
make[4]: *** [src/nvim/CMakeFiles/nvim.dir/fileio.c.o] Error 1
```

This appears to have been introduced by commit 07265d221fbd3b4fd16440caab49c70d14370079.

Not sure why this hasn't broken the build, I'm guessing this is because I am on
an ancient SLES box.

Platform details

```
$ uname -a
Linux cflcs01 3.0.101-0.47.71-default #1 SMP Thu Nov 12 12:22:22 UTC 2015 (b5b212e) x86_64 x86_64 x86_64 GNU/Linux

$ rpm -qif /usr/include/fcntl.h
Name        : glibc-devel                  Relocations: (not relocatable)
Version     : 2.11.3                            Vendor: SUSE LINUX Products GmbH, Nuernberg, Germany
Release     : 17.90.4                       Build Date: Thu 08 Oct 2015 01:15:31 PM CDT
Install Date: Sat 21 Nov 2015 09:54:50 AM CST      Build Host: sheep11
Group       : Development/Libraries/C and C++   Source RPM: glibc-2.11.3-17.90.4.src.rpm
Size        : 32759081                         License: BSD 3-Clause; GPL v2 or later; LGPL v2.1 or later
Signature   : RSA/8, Wed 14 Oct 2015 01:45:40 PM CDT, Key ID e3a5c360307e3d54
Packager    : https://www.suse.com/
URL         : http://www.gnu.org/software/libc/libc.html
Summary     : Include Files and Libraries Mandatory for Development
Description :
These libraries are needed to develop programs which use the standard C
library.
Distribution: SUSE Linux Enterprise 11
```

Build info

```
$ gcc --version
gcc (SUSE Linux) 4.3.4 [gcc-4_3-branch revision 152973]
Copyright (C) 2008 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

$ cmake --version
cmake version 3.4.20160116

CMake suite maintained and supported by Kitware (kitware.com/cmake).
```

**Comment by george-b** [comment]
I now see in my initial comment I am looking in the wrong fcntl.h file

```
grep NOFOLLOW /usr/include/asm-generic/fcntl.h
#ifndef O_NOFOLLOW
#define O_NOFOLLOW      00400000        /* don't follow links */
```

I can confirm I don't have this issue on a RHEL 6.7 box, I find this odd as I'm not touching my `$CPATH` or `$CFLAGS` in the example on the SLES system.

**Comment by sethjackson** [comment]
Probably because the `<fcntl.h>` that SLES has doesn't confirm to POSIX 2008.

**Comment by george-b** [comment]
The grep from my second comment is from SLES.

**Comment by sethjackson** [comment]
Hmm. Ok. That's really strange. Maybe that isn't the `<fcntl.h>` that gets included or something?

**Comment by george-b** [comment]
Maybe on both the RHEL and SLES boxes `cpp -v` lists `/usr/include`. 

```sh
$ for i in `find /usr/include -name fcntl.h | sort`; do echo $i && grep NOFOLLOW $i ; echo ; done
/usr/include/asm-arm/fcntl.h
#define O_NOFOLLOW      0100000 /* don't follow links */

/usr/include/asm-generic/fcntl.h
#ifndef O_NOFOLLOW
#define O_NOFOLLOW      00400000        /* don't follow links */

/usr/include/asm-ia64/fcntl.h

/usr/include/asm-microblaze/fcntl.h

/usr/include/asm-parisc/fcntl.h
#define O_NOFOLLOW      000000200 /* don't follow links */

/usr/include/asm-powerpc/fcntl.h
#define O_NOFOLLOW      0100000 /* don't follow links */

/usr/include/asm-s390/fcntl.h

/usr/include/asm-score/fcntl.h

/usr/include/asm-x86/fcntl.h

/usr/include/bits/fcntl.h
# define O_NOFOLLOW     0400000 /* Do not follow links.  */

/usr/include/fcntl.h
# define AT_SYMLINK_NOFOLLOW    0x100   /* Do not follow symbolic links.  */

/usr/include/linux/fcntl.h
#define AT_SYMLINK_NOFOLLOW     0x100   /* Do not follow symbolic links.  */

/usr/include/sys/fcntl.h
```

The file `/usr/include/asm-generic/fcntl.h` *should* be the first header file
found right?

Some googling has revealed the `-M` flag for GCC to me, so doing the following

```sh
$ CPATH=/usr/include:../.deps/usr/include:. gcc -M nvim/fileio.c | grep fcntl
  /usr/include/bits/wchar.h /usr/include/fcntl.h \
  /usr/include/bits/fcntl.h /usr/include/sys/types.h \

$ grep O_NOFOLLOW /usr/include/fcntl.h

$ grep O_NOFOLLOW /usr/include/bits/fcntl.h
# define O_NOFOLLOW     0400000 /* Do not follow links.  */
```

Seems to indicate we're not pulling in from `/usr/include/asm-generic/fcntl.h`
which is missing a leading zero. Is this what you were alluding to when you
mentioned not conforming to POSIX 2008?

Regardless the above is the same on the RHEL box where I don't see this issue, i.e. `gcc -M` also shows `/usr/include/bits/fcntl.h` so perhaps I'm barking up the wrong tree here.

**Comment by sethjackson** [comment]
> Is this what you were alluding to when you mentioned not conforming to POSIX 2008?

No POSIX 2008 says that `O_NOFOLLOW` should be in `<fcntl.h>`.

See [here](http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/fcntl.h.html).

So it looks like SLES has this definition but we are somehow not getting it. I don't currently have a SLES system to test but I might spin up a VM and check this out.

Anyone else have any ideas?

**Comment by Pyrohh** [comment]
Maybe it's behind a [feature test macro](https://www.gnu.org/software/libc/manual/html_node/Feature-Test-Macros.html)?

I'm not sure if glibc 2.11.3 supports it, but defining `_DEFAULT_SOURCE` at the top of `fileio.c` might work.

**Comment by Pyrohh** [comment]
If that doesn't work then `_POSIX_SOURCE=1` might.

**Comment by sethjackson** [comment]
> Maybe it's behind a feature test macro?

I hate `glibc`. This might be it. Forgot about this craziness.

@george-b can you just add `#define _DEFAULT_SOURCE` **before** the `<fcntl.h>` include and see if helps?

**Comment by sethjackson** [comment]
I just wanted to add that previously that `#ifndef O_NOFOLLOW` block would have probably done some really bad stuff here since it was defined to `0` previously and it's obviously not `0` on `glibc` from the above output.

**Comment by george-b** [comment]
I'm afraid neither of the above suggestions prevent the error I'm seeing.

> I hate glibc.

:+1:

**Comment by Pyrohh** [comment]
I think I found the problem, from glibc's [`open(2)` manual page](http://man7.org/linux/man-pages/man2/open.2.html):

```
The O_CLOEXEC, O_DIRECTORY, and O_NOFOLLOW flags are not specified in
POSIX.1-2001, but are specified in POSIX.1-2008.  Since glibc 2.12,
one can obtain their definitions by defining either _POSIX_C_SOURCE
with a value greater than or equal to 200809L or _XOPEN_SOURCE with a
value greater than or equal to 700.  In glibc 2.11 and earlier, one
obtains the definitions by defining _GNU_SOURCE.
```

So I guess you need `_GNU_SOURCE`.

**Comment by Pyrohh** [comment]
Oh, and you might also need the headers mentioned in that manual, too.

**Comment by george-b** [comment]
Yup defining `_GNU_SOURCE` does the trick. I end up having to do this in a couple of other files but I can now compile. Do you want to maintain this in tree? It would be nice to have neovim compile out of the box but I fully acknowledge this is an old platform and you probably wont have other users come across this.

```diff
diff --git a/src/nvim/fileio.c b/src/nvim/fileio.c
index 26376af..cdaa3a6 100644
--- a/src/nvim/fileio.c
+++ b/src/nvim/fileio.c
@@ -2,6 +2,8 @@
  * fileio.c: read from and write to a file
  */

+#define _GNU_SOURCE
+
 #include <assert.h>
 #include <errno.h>
 #include <stdbool.h>
diff --git a/src/nvim/memfile.c b/src/nvim/memfile.c
index 8cf5642..be6eb2b 100644
--- a/src/nvim/memfile.c
+++ b/src/nvim/memfile.c
@@ -35,6 +35,8 @@
 /// mf_trans_del()    may translate negative to positive block number
 /// mf_fullname()     make file name full path (use before first :cd)

+#define _GNU_SOURCE
+
 #include <assert.h>
 #include <inttypes.h>
 #include <limits.h>
diff --git a/src/nvim/shada.c b/src/nvim/shada.c
index dcdf219..db09112 100644
--- a/src/nvim/shada.c
+++ b/src/nvim/shada.c
@@ -1,3 +1,5 @@
+#define _GNU_SOURCE
+
 #include <stdlib.h>
 #include <stddef.h>
 #include <stdbool.h>
diff --git a/src/nvim/undo.c b/src/nvim/undo.c
index 6b60f95..6263c2a 100644
--- a/src/nvim/undo.c
+++ b/src/nvim/undo.c
@@ -72,6 +72,8 @@
 #define UH_MAGIC 0x18dade       /* value for uh_magic when in use */
 #define UE_MAGIC 0xabc123       /* value for ue_magic when in use */

+#define _GNU_SOURCE
+
 #include <assert.h>
 #include <inttypes.h>
 #include <limits.h>
```

**Comment by sethjackson** [comment]
I would not be adverse to making this work out of the box but only if we can avoid defining `_GNU_SOURCE` in the source files themselves.

I think we should be able to do this with CMake.

And if we don't maintain in tree for whatever reason you might be able do add `_GNU_SOURCE` in `local.mk`. See the example in `contrib/local.mk.example`.

**Comment by Pyrohh** [comment]
> Yup defining _GNU_SOURCE does the trick. I end up having to do this in a couple of other files but I can now compile. Do you want to maintain this in tree? It would be nice to have neovim compile out of the box but I fully acknowledge this is an old platform and you probably wont have other users come across this.

Given how old glibc 2.11.3 is, I'm skeptical of adding yet another build quirk. I wonder what others think. cc @jszakmeister @justinmk.

On some seldom used platforms, such as OpenBSD, where jemalloc doesn't work, we just have specific instructions on the wiki for what's needed in `local.mk`, as opposed to having something like this:
```cmake
if(CMAKE_SYSTEM_NAME STREQUAL "OpenBSD")
  disable_jemalloc()
endif()
```

> And if we don't maintain in tree for whatever reason you might be able do add _GNU_SOURCE in local.mk. See the example in contrib/local.mk.example.

We could add an example to `local.mk.example` and mention in the wiki instructions what users like @george-b need to add there, assuming we can do `-D_GNU_SOURCE` probject-wide and there won't be any conflicting declarations that arise.

**Comment by george-b** [comment]
I agree I don't want to clutter the source or the build system with workarounds for crufty old platforms. Indeed this probably goes against the mantra of neovim. It's trivial for me to work around this anyway.

**Comment by Pyrohh** [comment]
Sure, I'm glad the problem is fixed... somewhat.

I'll close this for now.

**Comment by justinmk** [comment]
libuv defines _GNU_SOURCE for several cases already: https://github.com/libuv/libuv/search?q=_GNU_SOURCE

And 2008 seems a bit late for a minimum target...
